### PR TITLE
CVE-2007-4559 Patch

### PR DESCRIPTION
**Pull request by TrellixVulnTeam**
# Patching CVE-2007-4559

Hi, we are security researchers from the Advanced Research Center at [Trellix](https://www.trellix.com). We have began a campaign to patch a widespread bug named CVE-2007-4559. CVE-2007-4559 is a 15 year old bug in the Python tarfile package. By using extract() or extractall() on a tarfile object without sanitizing input, a maliciously crafted .tar file could perform a directory path traversal attack. We found at least one unsantized extractall() in your codebase and are providing a patch for you via pull request. The patch essentially checks to see if all tarfile members will be extracted safely and throws an exception otherwise. We encourage you to use this patch or your own solution to secure against CVE-2007-4559. Further technical information about the vulnerability can be found in this [blog](https://www.trellix.com/en-us/about/newsroom/stories/research/tarfile-exploiting-the-world.html).

If you have further questions you may contact us through this projects lead researcher [Kasimir Schulz](mailto:kasimir.schulz@trellix.com).

### CODE REVIEW

File: src/compas/__init__.py [hunk 1]
```diff
@@ -303,7 +303,26 @@ def get_bunny(localstorage=None):
...
+                return prefix == abs_directory
+            
+            def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
+            
+                for member in tar.getmembers():
+                    member_path = os.path.join(path, member.name)
+                    if not is_within_directory(path, member_path):
+                        raise Exception("Attempted Path Traversal in Tar File")
+            
+                tar.extractall(path, members, numeric_owner)
```
Comment (chenkasirer) [inline src/compas/__init__.py]: @TrellixVulnTeam thanks for the fix!

However this call seems to be broken. According to the [docs ](https://docs.python.org/3.10/library/tarfile.html) `numeric_owner` is forced to be a kwarg.


```suggestion
                tar.extractall(path, members, numeric_owner=numeric_owner) 
```

**Review by chenkasirer — CHANGES_REQUESTED** [review]
@TrellixVulnTeam thanks for the fix!

The new call to `extractall` seems to be broken though.

File: src/compas/__init__.py [hunk 2]
```diff
@@ -303,7 +303,26 @@ def get_bunny(localstorage=None):
...
+            def is_within_directory(directory, target):
+                
+                abs_directory = os.path.abspath(directory)
+                abs_target = os.path.abspath(target)
+            
+                prefix = os.path.commonprefix([abs_directory, abs_target])
+                
+                return prefix == abs_directory
+            
+            def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
```
Comment (chenkasirer) [inline src/compas/__init__.py]:
```suggestion
            def safe_extract(tar, path=".", members=None, numeric_owner=False):
```

**Review by tomvanmele — APPROVED** [review]
LGTM

**Review by chenkasirer — APPROVED** [review]

**Pull request merged**